crates/zksync/core/src/vm/inspect.rs

```diff
@@ -329,7 +332,12 @@ where
 
     let mut storage: rHashMap<Address, rHashMap<rU256, StorageSlot>> = Default::default();
     let mut codes: rHashMap<Address, (B256, Bytecode)> = Default::default();
-    for (k, v) in &modified_storage {
+    // We skip nonce updates when should_update_nonce is false to avoid nonce mismatch
+    for (k, v) in modified_storage.iter().filter(|(k, _)| {
+        !(k.address() == &NONCE_HOLDER_ADDRESS &&
+            get_nonce_key(&initiator_address) == **k &&
+            !should_update_nonce)
+    }) {
```

Comment:
Let's not mix imperative and functional styles here. Either we can move the filter before the loop and `let filtered_modified_storage = ...` or simply use the `continue` statement within the loop with the same logic.